lib/testing/matrix.js

```diff
@@ -0,0 +1,172 @@
+    *
+    * @return {undefined}
+    */
+
+    /**
+    * Generate testing matrix, call execute method to run callback
+    *
+    * @param {string[]} elementsToSpecialize - key element we want to
+    *                    specialize on the matrix
+    * @param {GenerateMatrixCallback} callback - callback function
+    *                    who want to be call on execute method
+    *                    for the current generate matrix
+    *
+    * @returns {TestMatrix} Return current instance of Matrix object
+    */
+    generate(elementsToSpecialize, callback) {
+        this.elementsToSpecialize = elementsToSpecialize;
+        this.callback = callback;
+        return this;
+    }
+
+    /**
+    * Method to prevent an exception during the tests.
+    *
+    * @param {Object.<string, string[]>} exception - The keys of the hash table
+    is the element we want to do an exception, the value
+    *                      is the value we want to do an exception for the
+    *                      specific key.
+    *                      All key of the object must be validate with at least
+    *                      one value for execute an exception.
+    *                      ex: for a hashmap with 'auth' as key and 'v4' as
+    *                      value, if we found a matrix with this following
+    *                      paramater, it will call the special callback
+    * @argument {GenerateMatrixCallback} callbackException - Callback who the
+    *                       execute method will call when
+    *                       this exception is found.
+    *
+    * @returns {TestMatrix} Return current instance of Matrix object
+    */
+    if(exception, callbackException) {
+        const pair = {
+            key: exception,
+            callback: callbackException,
+        };
+        this.listOfExceptions.push(pair);
+
+        return this;
+    }
+
+    /**
+    * Return the parameters of the matrix into a Json format
+    *
+    * @returns {string} JSON format of parameters
+    */
+    serialize() {
+        return this.paramsToString;
+    }
+
+    /**
+    * Generate the matrix and call all callback enter on @generate and
+    * if methods. Must be run as the father matrix for running all the test
+    *
+    * @returns {undefined}
+    */
+    execute() {
+        /**
+        * Check if @generate method has been already called
+        */
+        if (this.callback) {
+            /**
+            * Generate matrix element by element, we call the @execute
+            * method until that we specialize all element send
+            * by generate method
+            */
+            if (this.elementsToSpecialize.length !== 0) {
+                /**
+                * Make copy for restore object in the end of recursive function
+                */
+                const elementToSpecialize = this.elementsToSpecialize[0];
+                const copyElement = this.params[elementToSpecialize];
```

Comment:
This is not actually a copy, but actually a reference to the value or array of possible values.
The simplest way to "fix" this is to rename into something along the lines of "originalValues".